src/main/scala/is/hail/io/vcf/ExportVCF.scala

```diff
@@ -230,73 +166,99 @@ object ExportVCF {
     }
   }
-
-  def apply[T >: Null](vkds: VariantSampleMatrix[Locus, Variant, T], path: String, append: Option[String] = None,
-    parallel: Boolean = false) {
-    val vas = vkds.vaSignature
-
-    val genotypeSignature = vkds.genotypeSignature
-
-    val (genotypeFormatField, genotypeFieldOrder) = genotypeSignature match {
-      case _: TGenotype =>
-        ("GT:AD:DP:GQ:PL", null)
-
-      case sig: TStruct =>
-        val fields = sig.fields
-        val formatFieldOrder: Array[Int] = sig.fieldIdx.get("GT") match {
-          case Some(i) => (i +: fields.filter(fd => fd.name != "GT").map(_.index)).toArray
-          case None => sig.fields.indices.toArray
-        }
-        val formatFieldString = formatFieldOrder.map(i => fields(i).name).mkString(":")
-
-        checkFormatSignature(sig)
-
-        (formatFieldString, formatFieldOrder)
-
-      case _ => fatal(s"Can only export to VCF with genotype signature of TGenotype or TStruct. Found `${ genotypeSignature }'.")
-    }
-
-    val infoSignature = vkds.vaSignature
-      .getAsOption[TStruct]("info")
-    val infoQuery: (Annotation) => Option[(Annotation, TStruct)] = infoSignature.map { struct =>
-      val (_, f) = vkds.queryVA("va.info")
-      (a: Annotation) => {
-        if (a == null)
-          None
-        else
-          Some((f(a), struct))
+  
+  def emitGenotype(sb: StringBuilder, formatFieldOrder: Array[Int], tg: TStruct, m: MemoryBuffer, offset: Long) {
+    formatFieldOrder.foreachBetween { j =>
+      val fIsDefined = tg.isFieldDefined(m, offset, j)
+      val fOffset = tg.loadField(m, offset, j)
+
+      tg.fields(j).typ match {
+        case TCall(_) =>
+          if (fIsDefined) {
+            strVCF(sb, TCall(), m, fOffset)
+          } else
+            sb.append("./.")
+        case it: TIterable =>
+          if (fIsDefined) {
+            val fLength = it.loadLength(m, fOffset)
+            iterableVCF(sb, it, m, fLength, fOffset)
+          } else
+            sb += '.'
+        case t =>
+          if (fIsDefined)
+            strVCF(sb, t, m, fOffset)
+          else
+            sb += '.'
       }
-    }.getOrElse((a: Annotation) => None)
+    }(sb += ':')
+  }
 
-    val hasSamples = vkds.nSamples > 0
+  def apply[RPK, RK, T >: Null](vkds0: VariantSampleMatrix[RPK, RK, T], path: String, append: Option[String] = None,
+    parallel: Boolean = false)(implicit tct: ClassTag[T]) {
+    
+    if (vkds0.sSignature != TString()) {
+      fatal(s"export_vcf requires s to have type TString, found ${vkds0.sSignature}")
+    }
+    
+    if (!vkds0.vSignature.isInstanceOf[TVariant])
+      fatal(s"export_vcf requires v to have type Variant, found ${vkds0.vSignature}")
+    
+    val vkds = vkds0.genotypeSignature match {
+      case TGenotype(_) =>
+        vkds0.annotateGenotypesExpr("g = {GT: Call(g.gt), AD: g.ad, DP: g.dp, GQ: g.gq, PL: g.pl}")
+          .copy2(genotypeSignature = TGenotype.representationWithVCFAttributes())
+      case t: TStruct => vkds0
```

Comment:
anything that was formerly of TGenotype will have the metadata attached from `TGenotype.representationWithVCFAttributes`. Will it be a problem if the other things don't?

Reply:
When we move exclusively to generic import of VCF and rip out TGenotype, then `TGenotype.representationWithVCFAttributes` will get ripped out too. This schema with the attributes will be created on import from a standard VCF by parsing the FORMAT lines in the header, and will then be reproduced on export using the same mechanism used generically here.